kkm/kkm_idt.c

```diff
@@ -215,6 +218,72 @@ int kkm_idt_descr_init(void)
 	*(uint64_t *)idt_entry->kx_global.va =
 		(uint64_t)kkm_switch_to_host_kernel;
 
+	/*
+	 * modify code to change pc relative instructions to direct address
+	 */
+	uint8_t *insn_addr = NULL;
+	uint32_t offset = -1;
+	uint64_t rip_value = -1;
+	uint64_t gs_relative_offset = -1;
+	uint8_t *relocated_address =
+		NULL; // this is where copied instruction is
+
+	insn_addr = (uint8_t *)kkm_syscall_label_1;
+	/* replace mov    %rsp,%gs:<offset>(%rip) with mov %rsp,%gs:<offset> */
+	if (insn_addr[0] == 0x65 && insn_addr[1] == 0x48 &&
+	    insn_addr[2] == 0x89 && insn_addr[3] == 0x25) {
+		offset = *(uint32_t *)(insn_addr + 4);
+		rip_value = (uint64_t)(insn_addr + 8);
+		gs_relative_offset = rip_value + offset;
+		relocated_address =
+			(uint8_t *)(kkm_syscall_label_1 - kkm_intr_entry_0 +
+				    idt_entry->idt_text_page0.va);
+		relocated_address[0] = 0x65;
+		relocated_address[1] = 0x48;
+		relocated_address[2] = 0x89;
+		relocated_address[3] = 0x24;
+		relocated_address[4] = 0x25;
+		*((uint32_t *)&relocated_address[5]) =
+			(uint32_t)gs_relative_offset;
```

Comment:
Do insn_addr and relocated_address refer to the same location in memory?
Or, do we have a copy of the page containing *insn_addr and we update the copy?  Where relocated_address refers to the copy.

Reply:
insn_addr is the kkm_syscall_label_? symbol address in kkm module.

relocated_address is in memory page allocated to copy some of the guest_exit code.